### PR TITLE
BMCJob and BMCTask spec definition

### PR DESCRIPTION
**Pull request by pokearu**
## Description
Spec and status for baseboard management job api. The key changes in the PR include,

1. Renamed PowerCycleJob as BMCJob. My idea was to have a more generic name to scope in different actions.
2. Added the spec for BMCJob. Example
```yaml
...
spec:
  tasks:
    - powerAction:
        state: "off"
    - bootAction:
        bootDevice: "pxe"
        persistent: "false"
        efiBoot: "false"
    - powerAction:
        state: "on"
```
Each `Task` may only have one type of action and the controller iterates and performs each of the requested actions.

3.  `.Status.Conditions` has a type `TaskCompleted` which indicates a task has completed and the `message` field would contain the success log. If a single task fails, condition `Failed` can be set for the whole job. 

### EDIT:
Based on discussions I have split the Job and Tasks to individual APIs. 

1. The BMCJob now has a list of tasks that are reconciled and used to create BMCTask API resources.
2. BMCTasks have their conditions and message to track individual task failures.
3. The BMCJob itself would fail when one of the tasks fail and would be updated with the name of the task resource that failed.

### CODE REVIEW

File: api/v1alpha1/baseboardmanagement_types.go
Comment (micahhausler) [inline api/v1alpha1/baseboardmanagement_types.go]: How do "status" and "soft" model a state?

File: api/v1alpha1/bmcjob_types.go
Comment (micahhausler) [inline api/v1alpha1/bmcjob_types.go:27]: Do you need an in-progress state?

File: api/v1alpha1/bmcjob_types.go
Comment (micahhausler) [inline api/v1alpha1/bmcjob_types.go]: Can you remove this note?

File: api/v1alpha1/bmcjob_types.go
Comment (micahhausler) [inline api/v1alpha1/bmcjob_types.go]: Can this be modeled as a slice so we can set an order?

File: api/v1alpha1/bmcjob_types.go
Comment (micahhausler) [inline api/v1alpha1/bmcjob_types.go]: I think we probably want to represent persistent state on the `BaseboardManagement` type, which would make this always ephemeral. In that case can we remove this flag?

File: api/v1alpha1/bmcjob_types.go
Comment (micahhausler) [inline api/v1alpha1/bmcjob_types.go]: Since the 'job' type is more imperative than declarative, it probably makes more sense for this to be a separate type from `PowerState` that is geared around actions than state. You will still want an "Off" and "On" action, but those are actions (verbs) rather than states (nouns) so you can include things like "ResetHard" and "ResetSoft". It doesn't make sense for the `BaseboardManagement` type to get placed in a `Reset` state.

File: api/v1alpha1/bmcjob_types.go
Comment (micahhausler) [inline api/v1alpha1/bmcjob_types.go]:
```suggestion
	EFIBoot bool `json:"efiBoot,omitempty"`
```

**Review by micahhausler — CHANGES_REQUESTED** [review]

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: I believe right now we wanted to hold off on modeling bootdevice from `BaseboardManagement` type? Since we were not sure if it would be a slice and since bmclib does not fetch boot device/order.

My intention here was to enable the job to be able to set persistent and one time bootdevice. 
Or is the suggestion to only set one time boot from job by default?

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: Right. That does make sense 👍  I can create a new type that is geared towards the actions and set that instead.

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: Sure! But at the moment bmclib would only enable us to set first boot device.
I can model it as a slice now and follow up with an issue on bmc-lib to fetch and set boot order 👍

File: api/v1alpha1/baseboardmanagement_types.go
Comment (pokearu) [reply api/v1alpha1/baseboardmanagement_types.go]: True. I was a bit iffy on this. I can rework them to `PowerAction` type which would make more sense.

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go:27]: I initially had this. But my idea was to add a new condition `TaskCompleted` when each task ran successfully. So we can track the progress.
We could have a `Running` condition, which I guess would be the first condition the job goes to?

File: config/crd/patches/webhook_in_bmcjobs.yaml
Comment (chrisdoherty4) [inline config/crd/patches/webhook_in_bmcjobs.yaml:5]: Why do we have patches for a new spec?

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [inline api/v1alpha1/bmcjob_types.go]: This needs a different type. We're conflating 2 different sets of state, that is a Job state and a Task state.

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [inline api/v1alpha1/bmcjob_types.go]:
```suggestion
type BootDeviceAction struct {
```

`BootAction` is too ambiguous.

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [reply api/v1alpha1/bmcjob_types.go]: Should this be just `Device`?

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [reply api/v1alpha1/bmcjob_types.go]: I think this comes back to the consumers desire. A BMC controller could, for example, create an action that performs a persistent boot set to bring about the desired boot state.

That said, I think we can only get by with only supporting one time boot and introduce this later. It wouldn't be a breaking API change so I suggest removing it in favor of temporary boot device set only.

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [inline api/v1alpha1/bmcjob_types.go]: We need commentary on whether we wait for the outcome of a task before proceeding with the next and whether we proceed to the next if the current fails. 

Those details are really quite important so we need to be very intentional. Happy to discuss more for rationale.

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: True. 
So my intention was to use this in the conditions for the `BMCJobStatus`. 
Would this be a different `Conditions` slice all together then? I was thinking to have task status as part of job status itself.

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [inline api/v1alpha1/bmcjob_types.go:24]: What are the possible types?

File: config/crd/patches/webhook_in_bmcjobs.yaml
Comment (pokearu) [reply config/crd/patches/webhook_in_bmcjobs.yaml:5]: I believe these files were auto-generated by `kubebuilder create api`?
@micahhausler Do you have more context on what they are used for?

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [reply api/v1alpha1/bmcjob_types.go]: We don't actually have a type for a task, its just a struct. If there are multiple tasks how does this work? It seems to be synonymous with `JobCompleted`.

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: Well that depends on if we want to represent it as a slice. I was thinking of `BootOrder`? And for now the controller would just set the first device on the `BootOrder`?
Unless we feel we don't want to do that now and limit to just one Device (first boot device). 

With that said, we could always add to the API in the future and give an option to set single device or entire boot order. Or a new action itself `BootOrderAction`.

File: config/crd/patches/webhook_in_bmcjobs.yaml
Comment (chrisdoherty4) [reply config/crd/patches/webhook_in_bmcjobs.yaml:5]: The probably are, but that doesn't matter much. Consolidate by regenerating everything ;).

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [reply api/v1alpha1/bmcjob_types.go]: What status does a task have and why would that be modeled on a job as a state? You could have `Tasks` and `CompletedTasks` moving tasks from one slice to the next so we know how far the job got but I'm not sure you'd want much more than that on the job type.

Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: My thought was, we would append to the `.Status.Conditions` slice. Like
```yaml
[ 
  {
    type: "TaskCompleted",
    message: "Power On task completed successfully"
  },
  {
    type: "TaskCompleted",
    message: "Boot Device PXE task completed successfully"
  },
  {
    type: "JobCompleted"
  }
]
```

This way we can track the individual tasks, so if a user wants to know what tasks have ran and even which failed.
That is why I kept it as a `BMCJobConditionType`, since tasks are part of a job.

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: Right. So if i understood correctly, we want to default to setting one time boot through the controller.
And later extend for persistent? Maybe even delegate that to a boot order setting action.
Sounds good to me 👍

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: Sure! I can add that. Right now my plan was

1. Wait for a task to complete before moving to next
2. If single task fails, we set the job as `Failed`
3. We only set `Completed` when all the tasks have completed successfully.

This does pose an argument where there can be cases in which we want to power off/reset even if the previous operation was not successful. Even maybe have retries? But I did not scope those for now.

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [reply api/v1alpha1/bmcjob_types.go]: Given this is wrapped up in a job I think we should model them as 2 separate actions actually. `BootOrderAction` and `OneTimeBootDeviceAction`. I suggest that because I think, given the discussion on the persistent flag, we're actually trying to model 2 distinct actions.

If we split into 2 distinct actions we can implement a `OneTimeBootDeviceAction` today and leave the persistent boot order construct to later because there are alternative ways of implementing that and we don't actually need it today. 

@micahhausler interested in your thoughts on the above too.

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [reply api/v1alpha1/bmcjob_types.go]: I see a few issues with that. (1) we're creating a dynamic number of conditions that could exist on a job; (2) consumers need to encode/cater for the dynamic state count in their logic (pushing details into a consumer is typically bad); (3) we're encoding information in a free form string about specific, finite tasks because all tasks are represented with a single type - thats different to opaque [error] messages.

On (3) that could be bad because a consumer may need to take corrective action and to do that they need to know how far the tasks went, to do that they need to inspect the free form string which, generally speaking, is a brittle design.

I can't decide if that's really bad but my gut tells me its not great - will need to bake on it. What I can say with confidence is that it increase the complexity of the API and I'm not sure its justifying the added complexity.

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: That sounds good to me 👍  By separating them we dont have to bake in logic later. And we can implement what we want now.

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: Right, that's a good point. 
My motivation was to have a way to track individual tasks without too much complexity. 
We could model it with a separate slice, so it does not interfere with the job conditions itself. Example

```go
type BMCJobStatus struct {
  TaskConditions []TaskCondition `json:"taskConditions,omitempty"`
}

type TaskCondition struct {
  Type TaskConditionType `json:"type"`
  Message string `json:"message,omitempty"`
}

type TaskConditionType string

const (
  TaskFailed TaskConditionType = "TaskFailed"
  PowerTaskCompleted TaskConditionType = "PowerTaskCompleted"
  ...
)
```

This way consumers can look at the job `conditions` with limited complexity and look at the `TaskConditions` (we could also call it `TaskStatuses` or `TaskResults`) to see what passed/failed.

File: api/v1alpha1/bmcjob_types.go
Comment (micahhausler) [inline api/v1alpha1/bmcjob_types.go]: Even though bmclib only supports a single device, changing this later to a list will involve a breaking API change. Can we model this as a slice, and only accept the first element?

**Review by micahhausler — CHANGES_REQUESTED** [review]

File: api/v1alpha1/bmcjob_types.go
Comment (pokearu) [reply api/v1alpha1/bmcjob_types.go]: Actually the idea here was to have it as a separate action. This action would only set one time boot. And not a boot order.


The discussion was going on in this thread
https://github.com/micahhausler/rufio-init/pull/3#discussion_r861348400

File: api/v1alpha1/baseboardmanagement_types.go
Comment (micahhausler) [inline api/v1alpha1/baseboardmanagement_types.go]: Is this too validation too strict? What if there are multiple NIC's, which one gets PXE booted, or same for Disks?

File: config/crd/patches/cainjection_in_bmctasks.yaml
Comment (micahhausler) [inline config/crd/patches/cainjection_in_bmctasks.yaml:6]: Why do we need certs here?

File: config/samples/bmc_v1alpha1_bmcjob.yaml
Comment (micahhausler) [inline config/samples/bmc_v1alpha1_bmcjob.yaml]: Can you update this to be that slice?
```suggestion
        devices: ["pxe"]
```

**Review by micahhausler — CHANGES_REQUESTED** [review]
This is looking great, just a few minor changes

File: config/crd/patches/cainjection_in_bmctasks.yaml
Comment (pokearu) [reply config/crd/patches/cainjection_in_bmctasks.yaml:6]: Actually these were auto generated by kubebuilder. Iam not a 100% but think if `certmanager` is enabled on the `kustomization.yaml` these might be needed?

These patches are currently [commented](https://github.com/micahhausler/rufio-init/blob/main/config/crd/kustomization.yaml#L16).

File: api/v1alpha1/baseboardmanagement_types.go
Comment (pokearu) [reply api/v1alpha1/baseboardmanagement_types.go]: I actually had these values to the currently available options on bmclib.

I think we have not worked out the scenario for multiple NICs and disks, would that be a separate BaseboardManagement object? Since it would have different IPs etc.

Comment (chrisdoherty4) [reply api/v1alpha1/baseboardmanagement_types.go]: I think it depends. We'd need to uniquely identify the device somehow and that identification would be dictated by the BMC interface spec, I'd imagine. I.e. I don't know that we know what data is required to model the right thing.

File: api/v1alpha1/bmcjob_types.go
Comment (chrisdoherty4) [reply api/v1alpha1/bmcjob_types.go:27]: Without running or in-progress you don't know if the job has started or not which is useful information.

Comment (chrisdoherty4) [reply api/v1alpha1/bmcjob_types.go:27]: I see you have `Running`, overlooked that. 

Same goes for tasks.

File: main.go
Comment (chrisdoherty4) [inline main.go]: For non-trivial errors, isolate the error check. Its far less cognitive load on maintainers.

```go
err := (&controllers.BMCJobReconciler{
	Client: mgr.GetClient(),
	Scheme: mgr.GetScheme(),
}).SetupWithManager(mgr)
if err != nil {
  // handle
}

File: api/v1alpha1/baseboardmanagement_types.go
Comment (pokearu) [reply api/v1alpha1/baseboardmanagement_types.go]: Yea I think we would need to scope that out a bit more. 
I could remove the validation for now so we don't limit ourselves 👍

Comment (micahhausler) [reply api/v1alpha1/baseboardmanagement_types.go]: Yea, lets remove this validation for now.

File: main.go
Comment (pokearu) [reply main.go]: Actually this was again auto generated by kubebuilder. I hadn't gotten to editing these files yet.
I can definitely change it up 👍

Comment (micahhausler) [reply main.go]: generally agree, but since this is auto-generated, so I don't care too much here

File: api/v1alpha1/bmctask_types.go
Comment (chrisdoherty4) [inline api/v1alpha1/bmctask_types.go:62]: Non-blocking: we need to clarify exactly what this does. I know it translates, under the hood, into passing `efiboot=true` in, for example, the IPMI payload but I think that behavior and result needs documenting at this level.

**Review by chrisdoherty4 — APPROVED** [review]

**Review by micahhausler — APPROVED** [review]
LGTM!

**Pull request merged**
